internal/k8s/watch.go

```diff
@@ -0,0 +1,144 @@
+package k8s
+
+import (
+	"log"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+// PodScrapeDetails stores the metrics endpoint details and metadata for a pod.
+type PodScrapeDetails struct {
+	Port      string
+	Path      string
+	PodName   string
+	Namespace string
+}
+
+// PodScrapeWatcher manages pod metrics and provides methods to handle updates and deletions.
+type PodScrapeWatcher struct {
+	PodMetricsEndpoints map[string]PodScrapeDetails
+	mu                  sync.Mutex
+
+	// Function variables for update and delete operations, to allow mocking during tests.
+	UpdatePodMetricsFunc func(*corev1.Pod)
+	DeletePodMetricsFunc func(*corev1.Pod)
+}
+
+const defaultResyncPeriod = 10 * time.Minute
+
+// NewPodScrapeWatcher initializes a new PodScrapeWatcher with default function implementations.
+func NewPodScrapeWatcher() *PodScrapeWatcher {
+	pw := &PodScrapeWatcher{
+		PodMetricsEndpoints: make(map[string]PodScrapeDetails),
+	}
+	pw.UpdatePodMetricsFunc = pw.UpdatePodMetrics
+	pw.DeletePodMetricsFunc = pw.DeletePodMetrics
+
+	return pw
+}
+
+// WatchPods starts the SharedInformer to monitor pod events and updates the metrics endpoints accordingly.
+func (pw *PodScrapeWatcher) WatchPods(clientset kubernetes.Interface, namespace string, labels map[string]string) {
+	factory := informers.NewSharedInformerFactoryWithOptions(
+		clientset,
+		defaultResyncPeriod,
+		informers.WithNamespace(namespace),
+		informers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+			opts.LabelSelector = metav1.FormatLabelSelector(&metav1.LabelSelector{MatchLabels: labels})
+		}),
+	)
+
+	podInformer := factory.Core().V1().Pods().Informer()
+
+	// Add event handlers for pod add/update/delete
+	if _, err := podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod, ok := obj.(*corev1.Pod)
+			if !ok {
+				log.Println("Error casting added object to Pod")
+				return
+			}
+			pw.UpdatePodMetricsFunc(pod)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			newPod, ok := newObj.(*corev1.Pod)
+			if !ok {
+				log.Println("Error casting updated object to Pod")
+				return
+			}
+			pw.UpdatePodMetricsFunc(newPod)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod, ok := obj.(*corev1.Pod)
+			if !ok {
+				log.Println("Error casting deleted object to Pod")
+				return
+			}
+			pw.DeletePodMetricsFunc(pod)
+		},
+	}); err != nil {
+		log.Fatalf("Failed to add event handler: %v", err)
+	}
+
+	// Start the informer
+	stopCh := make(chan struct{})
+	factory.Start(stopCh)
+	// Wait for the informer cache to sync
+	if !cache.WaitForCacheSync(stopCh, podInformer.HasSynced) {
+		close(stopCh) // Explicitly close the channel before exiting
+		log.Fatal("Failed to sync pod cache")
+	}
+
+	// Block until stopCh is closed
+	<-stopCh
```

Comment:
What is the significance of this?  If we removed this, would the behaviour change?  

I ask because I don't see anything waiting for `WatchPods` to finish, so I don't know what would go wrong if this just completed.  iiuc the informer would still be going in the background if we removed this, though I'm not certain about that

Reply:
So my knowledge is a bit limited here, yet afaik `factory.Start(stopCh)` natively starts the informer in a new routine, when the main thread finishes, the program ends and all `goroutines` are terminated as well, your statement is true in our case because of the running web server that keeps the main thread hanging, yet it kinda seems like best practice to keeps the `WatchPods` thread alive by blocking until something is sent on the `stopCh` channel, in our case would be graceful shutdowns.

See [client-go example](https://github.com/kubernetes/client-go/blob/43e9ea4bb53c86810ee4871bfd5c74258eff4ad0/examples/workqueue/main.go#L138)

Reply:
ok lgtm!